### DISCUSSION TITLE
autolink dueling algorithms

### DISCUSSION THREAD

**Discussion by agriffis** [micromark/micromark#49] [Q&A] [answered]
The README at https://github.com/micromark/micromark-extension-gfm-autolink-literal explains that GH has two algorithms for autolinking, and to get both parts, I would need to additionally use `mdast-util-gfm-autolink-literal`

I'm struggling to understand what I gain or lose by using only one half of the algorithm. In a quick test, using `micromark-extension-gfm-autolink-literal` is working fine:

```jsx
import {useMemo} from 'react'
import micromark from 'micromark'
import gfmAutolinkLiteralSyntax from 'micromark-extension-gfm-autolink-literal'
import gfmAutolinkLiteralHtml from 'micromark-extension-gfm-autolink-literal/html'

export const useMarkdown = source =>
  useMemo(
    () =>
      micromark(source, {
        extensions: [gfmAutolinkLiteralSyntax],
        htmlExtensions: [gfmAutolinkLiteralHtml],
      }),
    [source],
  )
```

Could someone shed some light on what I'm missing by leaving out `mdast-util-gfm-autolink-literal`?

**Comment by wooorm** [comment, marked as answer]
Some more info is here: https://github.com/micromark/micromark-extension-gfm-autolink-literal/pull/6.
I only recently realised that GH employs two widely differing algorithms, which they don’t document. That thread above shows how I figured it out. The main thing is that the “parsing” works if there are no “open” brackets. The “AST” one works if there is an open `[`.

**Reply by agriffis** [reply]
Thanks! That thread is helpful.

I'm still a bit uncertain how much it matters for my own use-case, which is a CMS without pre-existing expectations. The "parsing" approach seems to get to 90%, and the "AST" approach cleans up edge cases and parity with GitHub.

Mainly I want to avoid adding complexity that seems to [plague react-markdown with issues](https://github.com/remarkjs/react-markdown/issues/302). Whatever people put into the field, it shouldn't blow up.

(Also, I'm not really sure how `micromark-extension-gfm-autolink-literal` and `mdast-util-gfm-autolink-literal` come together in a single invocation of micromark, but I'll figure that out eventually...)

**Reply by wooorm** [reply]
That’s a pretty good summary: the first will work in 90% of cases, and you may choose whether that’s fine for you or not.

The react-markdown issues have more to do with a currently very naïve HTML implementation, which we’re again working on soon.

A small layer on micromark is https://github.com/syntax-tree/mdast-util-from-markdown. It can be given micromark and mdast extensions. The next steps would be https://github.com/syntax-tree/mdast-util-to-hast and https://github.com/syntax-tree/hast-util-to-html to get a string of HTML out.

**Reply by ChristianMurphy** [reply]
> > Mainly I want to avoid adding complexity that seems to plague react-markdown with issues. Whatever people put into the field, it shouldn't blow up.
>
> The react-markdown issues have more to do with a currently very naïve HTML implementation, which we’re again working on soon.

There is also a lighter version of react-markdown, which uses mdast under the hood, and can use the more complete html implementation (rehype) https://github.com/remarkjs/react-remark.
